src/libs/safe-path/src/safe_path_buf.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs::{self, File};
+use std::io::{Error, ErrorKind, Result};
+use std::ops::Deref;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::path::{Path, PathBuf};
+
+use crate::{open_by_path, safe_join};
+
+/// Safe version of `PathBuf` to protect from TOCTOU style of attacks.
+///
+/// There's a race window for attackers between time to validate a path and time to use the path.
+/// An attacker may maliciously change the path by symlinks to compose an attack.
+///
+/// The `/proc/self/fd/xxx` on linux is a symlink to the real target corresponding to the process's
+/// file descriptor `xxx`. And the symlink will be kept stable until the file descriptor has been
+/// closed. Combined with `O_PATH`, we could build a safe version of `PathBuf` by:
+/// - Generate a safe path from `root` and `path` by using [crate::safe_join()].
+/// - Open the safe path with O_PATH | O_CLOEXEC flags, say it's assigned `fd_num`.
+/// - Read the symlink target of `/proc/self/fd/fd_num`.
+/// - Compare the symlink target with the safe path, it's safe if these two paths equal.
+/// - Use the symlink target as a safe PathBuf.
+/// - Close the `fd_num` when dropping the `SafePathBuf` object.
+#[derive(Debug)]
+pub struct SafePathBuf {
+    file: File,
+    path: PathBuf,
+    target: PathBuf,
+}
+
+impl SafePathBuf {
+    /// Create a `SafePathBuf` from the `root` and an unsafe `path`.
+    ///
+    /// The `path` must be a subdirectory of `root`, otherwise error will be returned.
+    pub fn new<R: AsRef<Path>, U: AsRef<Path>>(root: R, path: U) -> Result<Self> {
+        let safe_path = safe_join(root, path)?;
+        Self::from_path(safe_path)
+    }
+
+    /// Create a `SafePathBuf` from an path.
+    ///
+    /// If the resolved value of `path` doesn't equal to `path`, an error will be returned.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
```

Comment:
This is exported for cases where we only want to lock down a path, instead of joining two paths. Output of `std::fs::canonicalize()` should also be a valid input for this function too.